### ISSUE TITLE
rewrite #!/usr/bin/env perl

### ISSUE THREAD

**Issue by miyagawa** [Perl-Toolchain-Gang/ExtUtils-MakeMaker#58]
(Somewhat related to #37 and #38?)

Is there any reason that ExtUtils::MakeMaker fixin() rewrites `#!/usr/bin/perl` or `#!perl` but NOT `#!/usr/bin/env perl`?

1) There are many reasons `#!/usr/bin/env perl` do not work when installed. Keeping that line in the shebang sounds wrong to me.

2) Yet, having `#!/usr/bin/env perl` is useful for scripts that you want to run from a git repository etc. Look what I do with cpanm bin scripts so that you can run from a git repo by cp, but also installable to the right perl path, by rewriting the shebang in Makefile.PL: https://github.com/miyagawa/cpanminus/blob/devel/Makefile.PL#L9

There might be a valid use case for allowing `#!/usr/bin/env perl` as stated in #38 - so it might make sense to "allow" it by leaving as is if `@INC` is configured as relocatable, but by default, it makes more sense to rewrite it upon install, the same way it does with `#!perl`.

**Comment by Grinnz** [comment]
There were some really good ideas near the end of this thread :+1: I just had another couple scripts installed with /usr/bin/env that then require manual intervention or gymnastics because they're not in my path.

**Comment by neilb** [comment]
Just ended up here, because `#!/usr/bin/env perl` wasn't "corrected" on install.

Like the proposal of this being fixed by default (as I expect most people happy & not surprised by that), with an option to turn it off.

@mohawk2, @bingos: how's your supply of tuits? :-)

**Comment by mohawk2** [comment]
I'll be pleased to write the probably small amount of code required, but I would like clear buy-in as to exactly what the rules are to be.

* Rewrite `#!perl` to `#!$PATH_TO_PERL`?
* Rewrite `#!/usr/bin/perl` to `#!$PATH_TO_PERL`?
* Rewrite `#!/usr/bin/env perl` to `#!$PATH_TO_PERL`?
* Have an option to rewrite any of the above to `#!/usr/bin/env perl`?
* Apply to all in `bin`? All in `scripts`?

**Comment by Grinnz** [comment]
The first two can be simplified to encompass the current "rule" of rewriting anything that has the initial command ending in `perl`.

**Comment by ap** [comment]
> Here's what I propose.
> 
> 1. Add a switch to turn fixin on or off (defaulting to on).
> 2. Change fixin so it will see #!/usr/bin/env perl as perl.
> 3. Allow the user to pass in a subroutine which rewrites the #! line.

That implies that `SHEBANG_FIX=0` would turn off all shebang translation. Is anyone looking for that feature? As far as I see, the conflict is only in what people intend to happen when they’ve used `#!/usr/bin/env perl`, whereas I can’t think of any reason anyone would ever want it disabled for `#!/path/ending/in/perl` – as evidenced by the lack of anyone ever complaining about it, as far as I can find.

So I don’t think the two should be coupled.

Also, this would up-and-change behaviour for a large swathe of CPAN, with fixups required downstream in any dist which *needs* the status quo.

So… I don’t expect anyone to care for my proposal, since “we patch all of CPAN every time we turn over a stone” seems en vogue nowadays, but as what feels like a futile gesture… here’s how this ends up looking when approached from a change management mindset:

It should be noted that this assumes, as it seems to be true, that people increasingly do not hand-maintain `Makefile.PL`. (Even those who have it checked into their source control tend to use some form of centrally maintained template or a boilerplate crib sheet or some such, I think.)

* Continue to *always* translate `#!/path/ending/in/perl` to `#!$^X`, i.e. no change in behaviour
* Add `SHEBANG_ENV_FIX` which rewrites `#!/usr/bin/env perl` when enabled
* Default it to off
* Ask the various authoring tools to emit `SHEBANG_ENV_FIX=1` by default (and also advertise it in the POD synopsis etc.), so newly created dists ship with it enabled by default, but old ones get the same treatment as before

(@neilb: river water quality and all that…?)

**Comment by kentfredric** [comment]
> *  Continue to always translate #!/path/ending/in/perl to #!$^X, i.e. no change in behaviour
> * Add SHEBANG_ENV_FIX which rewrites #!/usr/bin/env perl when enabled
> * Default it to off
> * Ask the various authoring tools to emit SHEBANG_ENV_FIX=1 by default (and also advertise it in the POD synopsis etc.), so newly created dists ship with it enabled by default, but old ones get the same treatment as before

This is more or less along what I'd hope for as well, though I don't mind if its an MM_OPT instead, I'd tack it along with my local:
```bash
PERL_MM_OPT+=" INSTALLDIRS=site"
```
That way packages can set a default for themselves if a user has no preference, but users (or vendors) with preference can override it.

Though I imagine if there's a way to control *which* SHEBANG will be used instead, I can also imagine that being *maybe* useful vendor side. That subject came up at least once in the last year, I just can't recall what the preferred direction of rewriting was.

If people are insistent that this kind of change be a default, we should at least give it adequate time to settle and test before we make it the default.

**Comment by neilb** [comment]
@ap: I agree, we should be careful not to break anything. I see at least three options for approaching this change:

 * Change the /usr/bin/env line by default. To do that we're have to be pretty damn confident that this wasn't going to break the distributions that have this line in scripts. We can easily identify how many there are, and what they are (I'd be curious how many authors might want /usr/bin/env to stay, and why).
 * Your proposed change, which means any newly shipped dists are affected. The danger is that someone may ship a dist, unaware of the change, and thus get behaviour they didn't want
 * Make the additional change opt-in. Eg I could use a key in Makefile.PL. At some point in the future we could considering switching that to opt-out.

We also have to remember that this is not just about CPAN. Even if we reassure ourselves that automatically making this change won't break anything on CPAN, it could still break things on DarkPAN.

So while I personally would be happy for this to be automatic, I think it should be opt-in.

**Comment by Grinnz** [comment]
I'll just note that unless this is done for users by default (or with some way to default it on in Makefile.PL), it's completely pointless as authors will continue to not be able to ship `#!/usr/bin/env perl`. This is also somewhat true regardless of any change, because there will always be users with an old toolchain, though a configure_requires on newer EUMM for dists using this feature could partially solve that.

**Comment by kentfredric** [comment]
> We can easily identify how many there are, and what they are (I'd be curious how many authors might want /usr/bin/env to stay, and why).

If recent experience is anything to go by, no, we can't.

We can only at best identify distributions that touch this code *on CPAN*, and then, we're likely to miss something until the day we deploy it, and then discover that package hasn't seen a maintainer in 5 years and people are relying on it.

It says *nothing* about what we're affecting in the darkpan, or more significantly, the grey-pans, the software that is widely used in opensource, but isn't published to CPAN ( not specific to this problem, but examples of software in the grey-pan domain are automake, autoconf, intltool, make, go ( go uses Perl in its tests ) ... for instance, and all of these cited examples had breakages under Perl 5.26 )

Its not safe to make any solid conclusions about the reality we're affecting by CPAN observations alone. It is a useful datapoint, but not enough in itself.

While I don't think there's a high chance of something being severely broken by such a change, I am *regularly* surprised by "Things that couldn't be imagined to cause problems", causing problems.

Like for example, I never imagined that injecting a `use lib q[.];` into a Makefile.PL while retaining perfectly valid Perl code would break anything. [But then it did, and created a cascade of things breaking](https://rt.cpan.org/Ticket/Display.html?id=122166)

Its safer to assume I'm stupid and then let tests argue my case. 

> We also have to remember that this is not just about CPAN. Even if we reassure ourselves that
> automatically making this change won't break anything on CPAN, it could still break things on DarkPAN.

> So while I personally would be happy for this to be automatic, I think it should be opt-in.

So yes, a thousand times this. The way people talk about darkpan sometimes is they make it sound like a nebulous boogie monster which is mostly an imagined problem that's not our job to worry about, that is just there as a dark story to keep you scared.

But its so much worse than just a scary story.

**Comment by ap** [comment]
@neilb:

> I'd be curious how many authors might want /usr/bin/env to stay, and why

At least one use case was [mentioned upthread](#issuecomment-17455661) by mst:

> I do this regularly for pure perl scripts that I want bundled into a relocatable local::lib.

**Comment by Grinnz** [comment]
> At least one use case was mentioned upthread by mst:

He also said that an option to keep the /usr/bin/env line is perfectly fine for such an use case, though.

**Comment by hartzell** [comment]
I have a project that includes a layer of scripts that do things and a second layer of scripts that call scripts from the first layer.

I'd like the installed scripts to use the perl interpreter that was used to build/test/install the package, which is to say I'd like their shebang lines rewritten.

I'd like to be able to call scripts in the source tree and have them work.  I realize that (as [@karenetheridge pointed out](https://github.com/Perl-Toolchain-Gang/ExtUtils-MakeMaker/issues/58#issuecomment-17199427)) I can call the first level scripts as `perl bin/first-level-script.pl` but that doesn't help when I call a second level script which runs `bin/first-level-script.pl` (via IPC::Cmd), I'd have to make the second level scripts run `$^X bin/first-level-script.pl` explicitly and explain why so that the next person to touch the code doesn't trip over it.

It feels cleaner to just call the scripts and arrange for them to work in each of their settings.

I ended up using the `MakeMaker::Awesome` dzil plugin (*the docs are as awesome as the plugin!*) to add a custom `MY` package to my `Makefile.PL` that replaces the call to MakeMaker's `fixin` routine with a simple call to `perl -pi -e s|...|...|`.

I'd appreciate any feedback on:

- the Dist::Zilla-isms;
- the ExtUtils-MakeMaker-isms; and/or
- the general approach to having a hierarchy of scripts work and be
  testable in the source tree and when installed.

Here's the plugin that I ended up with:

```perl
package inc::ShebangHackery;
use Moose;

extends 'Dist::Zilla::Plugin::MakeMaker::Awesome';

around _build_MakeFile_PL_template => sub {
    my $orig = shift;
    my $self = shift;

    # grab the original template
    my $template = $self->$orig(@_);

    # tack our MY class onto the end of it
    $template .= <<'TEMPLATE';
package MY;
use Cwd;

# The real FIXIN won't fix the shebang line we need, so Do It Ourselves.  Where
# the Makefile used to call $(FIXIN) it now uses perl's in-place editing
# capability to rewrite our shebang.  For background, see
# https://github.com/Perl-Toolchain-Gang/ExtUtils-MakeMaker/issues/58
sub installbin {
    my $inherited = shift->SUPER::installbin(@_);
    my $fixin = qq($^X -pi -e "s|^#!/usr/bin/env perl|#!$^X|");
    $inherited =~ s/\$\(FIXIN\)/$fixin/g;
    $inherited;
}
TEMPLATE

    return $template;
};

__PACKAGE__->meta->make_immutable;
```

**Comment by Grinnz** [comment]
In my opinion you should still use $^X to call the other script, so that it works deterministically regardless of how the user called the first one. The trickier part is passing on any `@INC` modifications to the second call (for example if a user does `/path/to/perl -I/path/to/lib /path/to/script`).

**Comment by mohawk2** [comment]
That last one is what `App::Prove` currently does, right? Maybe that needs extracting into a separately-usable module, maybe something like `lib::inherit::callperl(@argv)`?

**Comment by sdondley** [comment]
So this is merged now? Just so I'm clear, this means a distribution developer can leave `#!/usr/bin/env perl` in a script and it will be changed to an appropriate path when installed by the end user without the developer having to do anything?

**Comment by Grinnz** [comment]
No, nothing has been done yet.

**Comment by sdondley** [comment]
Oh, OK, I mistook the purple Merged symbols for this project but they are for others.

**Comment by toddr** [comment]
I've read this thread twice just now and I'm still not sure I understand where everyone's at so let me see if I can summarize and get some people to yell at me and tell me I'm wrong :D
 
1. The decision on what the shebang be should be in the power of the person installing the script, not the CPAN author. In most if not all cases, the CPAN author doesn't actually know what every person doing the install wants.
2. People with a re-locatable perl need a way for the shebang to not be `#!$^X`. 

**It seems like the discussed solution would be to:**
1. Add #!/usr/bin/env perl to the patterns that leads to a #! rewrite.
2. Continue to default the re-write to $^X
3. Provide a means to alter the rewrite value to something else. Most people want `/usr/bin/env perl` but others might want something else. `SHEBANG_FIX=1` produces `/usr/bin/env perl` or SHEBANG_FIX=/my/custom/path/to/whatever provides a way for them to be explicit about what they want it to be.

Right now everyone's unhappy. Some CPAN distros rewrite to `$^X` which angers the relocatable perl folks. Some CPAN distros use `/usr/bin/env perl` which frustrates the distros that don't want to have to be dependent on $PATH to get the perl script they explicitly just called to behave correctly.

Is anyone actually against this proposed fix?

**Comment by toddr** [comment]
> I am. I am actually in favor of each one of the points in the proposal, but I am against the proposal as a whole, because it is incomplete. I imagine everyone else on the thread so far who voiced concerns about breakage feels the same way.

Many of the comments above were answered and this ticketing system doesn't really allow for threading so it's not clear what's "unanswered".  Could you please be clear what additional change / improvement would be needed?

**Comment by toddr** [comment]
> But missing from it is any mechanism to minimise disruption resulting from erasing the hitherto distinction between `#!perl` and `#!/usr/bin/env perl`.

Sorry, I missed this part of your reply. 

It's my opinion that the decision on what the shebang should be should be in the power of the person installing the script, not the CPAN author. In most if not all cases, the CPAN author doesn't actually know what every person doing the install wants.

So my question to you: Under what circumstances would the user or distro maintainer actually benefit from some of the shebangs in a distro being `$^X` and some being `/usr/bin/env perl`? I can see a use for both but not in the same perl install.

If you come up with one, is there any reason the distro maintainer couldn't just patch that in their edge case? We're already planning to provide a command line option to specify the shebang in every module install.

**Comment by toddr** [comment]
What about support for an environment variable which tells the fixins the installer wants `/usr/bin/env perl` re-written to the perl used with Makefile.PL?

**Comment by Grinnz** [comment]
The vast majority of installers will not know they want this - shebang fix needs to "just work", but certainly there should be a way to skip rewriting, though my uninformed opinion is there should be an option to rewrite all shebangs to `/usr/bin/env perl` since many distros don't use this to begin with as rewriting is currently broken for it.

**Comment by toddr** [comment]
The big "But what about" seems to be relocatable perl. That's a %Config value. Maybe relocatable perl could preserve /usr/bin/env when that is the given perl?

**Comment by toddr** [comment]
> The vast majority of installers will not know they want this.

I agree but this ticket doesn't seem to be going anywhere so I'm looking for a compromise. Perhaps cpan.pm could set the environment var, etc and we could move forward.
